### PR TITLE
SDN-4599: Configurable network diagnostics pod placement

### PR DESCRIPTION
**Pull request by kyrtapz**
Enhancement: https://github.com/openshift/enhancements/pull/1585

### CODE REVIEW

**Comment by kyrtapz** [comment]
`verify-crd-schema` is failing on:
```
error in 0000_10_config-operator_01_network-CustomNoUpgrade.crd.yaml: NoMaps: crd/networks.config.openshift.io version/v1 field/^.spec.networkDiagnostics.targetPlacement.nodeSelector may not be a map
error in 0000_10_config-operator_01_network-CustomNoUpgrade.crd.yaml: NoMaps: crd/networks.config.openshift.io version/v1 field/^.spec.networkDiagnostics.sourcePlacement.nodeSelector may not be a map
error in 0000_10_config-operator_01_network-TechPreviewNoUpgrade.crd.yaml: NoMaps: crd/networks.config.openshift.io version/v1 field/^.spec.networkDiagnostics.sourcePlacement.nodeSelector may not be a map
error in 0000_10_config-operator_01_network-TechPreviewNoUpgrade.crd.yaml: NoMaps: crd/networks.config.openshift.io version/v1 field/^.spec.networkDiagnostics.targetPlacement.nodeSelector may not be a map 
```
Since  the `nodeSelector` is essentially label selector I think that it should be a map.

File: config/v1/types_network.go [hunk 1]
```diff
@@ -92,14 +106,15 @@ type NetworkStatus struct {
 
 	// conditions represents the observations of a network.config current state.
 	// Known .status.conditions.type are: "NetworkTypeMigrationInProgress", "NetworkTypeMigrationMTUReady",
-	// "NetworkTypeMigrationTargetCNIAvailable", "NetworkTypeMigrationTargetCNIInUse"
-	// and "NetworkTypeMigrationOriginalCNIPurged"
+	// "NetworkTypeMigrationTargetCNIAvailable", "NetworkTypeMigrationTargetCNIInUse",
+	//  "NetworkTypeMigrationOriginalCNIPurged" and "NetworkDiagnosticsAvailable"
```
Comment (trozet) [inline config/v1/types_network.go:110]: nit: is there an extra space here after // ?

**Review by trozet — APPROVED** [review]
/lgtm

File: config/v1/types_network.go [hunk 2]
```diff
@@ -70,6 +74,16 @@ type NetworkSpec struct {
 	// installed.
 	// +kubebuilder:validation:Pattern=`^([0-9]{1,4}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])-([0-9]{1,4}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])$`
 	ServiceNodePortRange string `json:"serviceNodePortRange,omitempty"`
+
+	// networkDiagnostics defines network diagnostics configuration.
+	//
+	// Cannot be specified when spec.disableNetworkDiagnostics is set to true in network.operator.openshift.io.
```
Comment (deads2k) [inline config/v1/types_network.go:80]: This looks backwards.  The configuration in config.openshift.io should be authoritative.

File: config/v1/types_network.go [hunk 3]
```diff
@@ -197,3 +212,62 @@ type MTUMigrationValues struct {
...
+	//
+	// +optional
+	TargetPlacement NetworkDiagnosticsNodePlacement `json:"targetPlacement"`
+}
+
+// NetworkDiagnosticsNodePlacement defines node scheduling configuration network diagnostics components
+type NetworkDiagnosticsNodePlacement struct {
+	// nodeSelector is the node selector applied to network diagnostics components
+	//
+	// By default this is set to `kubernetes.io/os: linux`
```
Comment (deads2k) [inline config/v1/types_network.go:258]: is it set or does empty mean this?

File: config/v1/types_network.go [hunk 4]
```diff
@@ -197,3 +213,86 @@ type MTUMigrationValues struct {
...
 }
+
+// NetworkDiagnosticsMode is an enumeration of the available network diagnostics modes
+// Valid values are "All", "Disabled".
+// +kubebuilder:validation:Enum:=All;Disabled
+type NetworkDiagnosticsMode string
+
+const (
+	// NetworkDiagnosticsAll means that all network diagnostics checks are enabled
+	NetworkDiagnosticsAll NetworkDiagnosticsMode = "All"
```
Comment (deads2k) [inline config/v1/types_network.go:224]: need to add an empty to allow empty.

File: config/v1/types_network.go [hunk 5]
```diff
@@ -197,3 +213,86 @@ type MTUMigrationValues struct {
 	// +optional
 	From *uint32 `json:"from,omitempty"`
 }
+
+// NetworkDiagnosticsMode is an enumeration of the available network diagnostics modes
+// Valid values are "All", "Disabled".
+// +kubebuilder:validation:Enum:=All;Disabled
```
Comment (deads2k) [inline config/v1/types_network.go:219]: need to add an empty to allow empty.

File: config/v1/tests/networks.config.openshift.io/NetworkDiagnosticsConfig.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+apiVersion: apiextensions.k8s.io/v1 # Hack because controller-gen complains if we don't have this
+name: "Network"
+crdName: networks.config.openshift.io
+featureGate: NetworkDiagnosticsConfig
+tests:
+  onCreate:
```
Comment (deads2k) [inline config/v1/tests/networks.config.openshift.io/NetworkDiagnosticsConfig.yaml:6]: Add a test to be sure that an empty object is creatable.

File: config/v1/types_network.go (hunk 2 above)
Comment (kyrtapz) [reply config/v1/types_network.go]: Changed it so networkDiagnostics takes precedence over disableNetworkDiagnostics if specified.

File: config/v1/types_network.go [hunk 7]
```diff
@@ -197,3 +212,62 @@ type MTUMigrationValues struct {
...
+
+// NetworkDiagnostics defines network diagnostics configuration
+
+type NetworkDiagnostics struct {
+	// mode controls the network diagnostics mode
+	//
+	// By default the value is set to All.
+	//
+	// +optional
+	// +kubebuilder:default=All
```
[the remark this answers is not in the document]
Comment (kyrtapz) [reply config/v1/types_network.go]: Thanks, good to know!
Done.

File: config/v1/types_network.go [hunk 8]
```diff
@@ -197,3 +212,62 @@ type MTUMigrationValues struct {
...
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector"`
+
+	// tolerations is a list of tolerations applied to network diagnostics components
+	//
+	// For SourcePlacement, this is set to an empty list by default.
+	//
+	// For TargetPlacement, this is set to `- operator: "Exists"` by default.
+	// It means that it tolerates all taints.
+	//
```
[the remark this answers is not in the document]
Comment (kyrtapz) [reply config/v1/types_network.go]: Created two types as suggested.

File: config/v1/types_network.go (hunk 4 above)
Comment (kyrtapz) [reply config/v1/types_network.go:227]: Done.

File: config/v1/tests/networks.config.openshift.io/NetworkDiagnosticsConfig.yaml (hunk 6 above)
Comment (kyrtapz) [reply config/v1/tests/networks.config.openshift.io/NetworkDiagnosticsConfig.yaml:6]: Done.

File: config/v1/types_network.go (hunk 3 above)
Comment (kyrtapz) [reply config/v1/types_network.go]: Added a clearer explanation.
